packages/story-editor/src/app/mediaUpload/provider.js

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+
+import { useRef } from '@web-stories-wp/react';
+import PropTypes from 'prop-types';
+/**
+ * Internal dependencies
+ */
+import { useConfig } from '..';
+import Context from './context';
+
+function OpenMediaUploadHandler({ onOpenMediaUpload, ...props }) {
+  /**
+   * `open` needs to be stored in a ref so that it does not cause endless
+   * re-renders for any children.
+   */
+  const openRef = useRef(open);
+  openRef.current = onOpenMediaUpload;
+
+  return (
+    <Context.Provider
+      {...props}
+      value={{ onOpenMediaUpload: openRef.current }}
+    />
+  );
+}
+OpenMediaUploadHandler.propTypes = {
+  onOpenMediaUpload: PropTypes.func.isRequired,
+};
+
+/**
+ * The media upload provider.
+ *
+ * The `MediaUpload` component is passed into the application from the
+ * config. There is not a way to access the `open` function outside
+ * of the `render` prop. This provider gives a way to access the `open`
+ * function for use anywhere in the application, passed through context.
+ *
+ * NOTE: Every time the wording and functionality of the modal
+ * needs to change, the provider must be re-rendered with those arguments!!!
+ *
+ * Ex:
+ * ```js
+ * return (
+ *  <MediaUploadProvider
+ *    title={__('Select media', 'web-stories')}
+ *    buttonInsertText={__('Insert media', 'web-stories')}
+ *  >
+ *    {...}
+ *  </MediaUploadProvider>
+ * )
+ * ```
+ * vs
+ * ```js
+ * return (
+ *  <MediaUploadProvider
+ *    title={__('Replace media', 'web-stories')}
+ *    buttonInsertText={__('Put in the media', 'web-stories')}
+ *  >
+ *    {...}
+ *  </MediaUploadProvider>
+ * )
+ * ```
+ *
+ * @param {Object} props Provider props
+ * @param {Node} props.children The children
+ * @return {Node} The wrapped children
+ */
+function MediaUploadProvider({ children, ...props }) {
+  const { MediaUpload } = useConfig();
+
+  return (
+    <MediaUpload
+      {...props}
+      render={(open) => (
+        /* the only way to access `open` is through the render prop of `MediaUpload` */
+        <OpenMediaUploadHandler onOpenMediaUpload={open}>
+          {children}
+        </OpenMediaUploadHandler>
+      )}
```

Comment:
maybe memoize since this is held at a high level:
```js
const renderMediaProvider = useCallback((open) => (
        /* the only way to access `open` is through the render prop of `MediaUpload` */
        <OpenMediaUploadHandler onOpenMediaUpload={open}>
          {children}
        </OpenMediaUploadHandler>
), []);
// ...
render={renderMediaProvider}
```

Reply:
I see the problem and why we are doing this. However, where `render` is supposed to render the button, it is weird to render the entire app inside `render` to make `open` available in the required component. 

Since `MediaUpload` can render the replace background icon, I think it's a matter of how the code is structured in our core story editor. So maybe if we think in a different direction of refactoring the existing code to use the `MediaUpload` component for rendering that icon as it should be, instead of hacking the `open` callback, it would make more sense.

Reply:
> refactoring the existing code to use the MediaUpload component for rendering that icon

Hmm, I haven't been in the trenches here but I think I broadly agree with this suggestion.

The current approach is pretty clever to workaround `MediaUpload`'s use of a [render prop](https://reactjs.org/docs/render-props.html). 

However, instead of adding `OpenMediaUploadHandler` and `MediaUploadProvider`, could we instead extend the `useQuickActions` API? 

Add an alternative to the `onClick` property -- maybe a `wrapper` property that decorates an existing component or definition? Then in `useQuickActions.js`, we could have something like:

```js
{
  Icon: PictureSwap,
  label: ACTIONS.REPLACE_BACKGROUND_MEDIA.text,
  onClickWrapper: (ToBeWrapped) => {
    return <MediaUpload
      render={(open) => <ToBeWrapped onClick={open}>},
      // having these properties below keeps media-related code local to its usage
      title={__('Replace media', 'web-stories')},
      onSelect={handleMediaSelect},
      ...
    />;
  },
  ...actionMenuProps,
},
```

And then in `menuItem.js`, something like:

```js
const ButtonWithoutClick = () => 
  <Button
    ref={itemRef}
    aria-label={itemLabel}
    disabled={disabled}
    onFocus={onFocus}
    {...menuItemProps}
  >
    {textContent}
  </Button>;

if (onClick) { // existing behavior
  return <ButtonWithoutClick onClick={handleClick}>
} else if (onClickWrapper) { // new behavior
  return onClickWrapper(ButtonWithoutClick);
}
```

It'd be nice to keep new code more local and avoid adding a root-level provider (especially one for each use of `MediaUpload`).